File: Week_07/130/surrounded_regions.py
from typing import List


class Solution:
    def solve(self, board: List[List[str]]) -> None:
        if not board or not board[0]:
            return
        m, n = len(board), len(board[0])

        def dfs(x: int, y: int):
            if x < 0 or x >= m or y < 0 or y >= n or board[x][y] != "O":
                return
            board[x][y] = "B"
            dfs(x - 1, y)
            dfs(x + 1, y)
            dfs(x, y - 1)
            dfs(x, y + 1)

        for i in range(m):
            for j in range(n):
                if board[i][j] == "O" and i == 0 or i == m - 1 or j == 0 or j == n - 1:
                    dfs(i, j)
        for i in range(m):
            for j in range(n):
                if board[i][j] == "O":
                    board[i][j] = 'X'
                elif board[i][j] == "B":
                    board[i][j] = "O"

File: Week_07/130/surrounded_regions.go
package main

func solve(board [][]byte) {
	if len(board) == 0 || len(board[0]) == 0 {
		return
	}

	var dfs func(x, y int)
	m, n := len(board), len(board[0])

	dfs = func(x, y int) {
		if x < 0 || x >= m || y < 0 || y >= n || board[x][y] != 'O' {
			return
		}
		board[x][y] = 'B'
		dfs(x-1, y)
		dfs(x+1, y)
		dfs(x, y-1)
		dfs(x, y+1)
	}

	for i := 0; i < m; i++ {
		for j := 0; j < n; j++ {
			if board[i][j] == 'O' && i == 0 || i == m-1 || j == 0 || j == n-1 {
				dfs(i, j)
			}
		}
	}
	for i := 0; i < m; i++ {
		for j := 0; j < n; j++ {
			if board[i][j] == 'O' {
				board[i][j] = 'X'
			} else if board[i][j] == 'B' {
				board[i][j] = 'O'
			}
		}
	}
}
